### PR TITLE
feat: java Kafka To BQ via Dstream template

### PR DESCRIPTION
**Pull request by chakresh84**
https://github.com/GoogleCloudPlatform/dataproc-templates/issues/897 
New java template  KafkaToBQDstream (Kafka To BQ via Dstream) for Dataproc

### CODE REVIEW

File: java/pom.xml [hunk 1]
```diff
@@ -322,6 +322,13 @@
       <version>2.1.0</version>
     </dependency>
 
+    <!-- Spark kafka Direct stream-->
+    <dependency>
+      <groupId>org.apache.spark</groupId>
+      <artifactId>spark-streaming-kafka-0-10_2.12</artifactId>
```
Comment (vanshaj-bhatia) [inline java/pom.xml:328]: Replace `2.12` with `${scala.binary.version}`

File: java/src/main/resources/template.properties [hunk 2]
```diff
@@ -457,3 +457,7 @@ text.bigquery.output.dataset=
 text.bigquery.output.table=
 text.bigquery.output.mode=
 text.bigquery.temp.bucket=
+
+# Kafka TO BQ Dstream
+kafka.bq.consumer.group.id=group-id1
```
Comment (vanshaj-bhatia) [inline java/src/main/resources/template.properties:462]: Remove `group-id1`, it shouldn't be a default value if template property `kafka.bq.consumer.group.id` is not provided

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstream.java [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+              ((CanCommitOffsets) stream.inputDStream()).commitAsync(offsetRanges);
+            });
+
+    ssc.start();
+    ssc.awaitTerminationOrTimeout(kafkaAwaitTerminationTimeout);
+  }
+
+  @Override
+  public void validateInput() {
+    if (StringUtils.isAllBlank(kafkaBootstrapServers)
```
Comment (vanshaj-bhatia) [inline java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstream.java:153]: let us make `kafkaGroupId`(KAFKA_BQ_CONSUMER_GROUP_ID) as required input.

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstream.java [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+    Map<String, Object> kafkaParams = new HashMap<>();
+    kafkaParams.put("bootstrap.servers", kafkaBootstrapServers);
+    kafkaParams.put("key.deserializer", StringDeserializer.class);
+    kafkaParams.put("value.deserializer", StringDeserializer.class);
+    kafkaParams.put("group.id", kafkaGroupId);
+    kafkaParams.put("auto.offset.reset", kafkaStartingOffsets);
+    kafkaParams.put("enable.auto.commit", false);
+
+    Collection<String> topics = Collections.singletonList(kafkaTopic);
```
Comment (vanshaj-bhatia) [inline java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstream.java:98]: Can kafkaTopic include multiple topic list like: t1,t2,t3 ?

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstream.java [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+          KAFKA_BQ_TABLE,
+          KAFKA_BQ_TEMP_GCS_BUCKET);
+      throw new IllegalArgumentException(
+          "Required parameters for KafkaToBQDstream not passed. "
+              + "Set mandatory parameter for KafkaToBQDstream template "
+              + "in resources/conf/template.properties file.");
+    }
+
+    LOGGER.info(
+        "Starting Kafka to BQ spark job with following parameters:"
```
Comment (vanshaj-bhatia) [inline java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstream.java:174]: `Kafka to BQ` -> `Kafka to BQ via DStream`

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstream.java [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+            + "2. {}:{} "
+            + "3. {}:{} "
+            + "4. {},{} "
+            + "5. {},{} "
+            + "6. {},{} "
+            + "7. {},{} "
+            + "8. {},{} ",
+        KAFKA_BOOTSTRAP_SERVERS,
+        kafkaBootstrapServers,
+        KAFKA_TOPIC,
```
Comment (vanshaj-bhatia) [inline java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstream.java:185]: Let us log the kafka group id here as well,.

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md [hunk 7]
```diff
@@ -237,4 +237,82 @@ bin/start.sh \
...
+# Offset to start reading from. Accepted values: "earliest", "latest" 
+kafka.bq.starting.offset=<kafka-starting-offset>
+
+# Waits for specified time in ms before termination of stream
+kafka.bq.await.termination.timeout=<stream-await-termination-timeout>
+
+# Ouptut mode for writing data. Accepted values: 'overwrite', 'append', 'ignore', 'error', 'errorifexists', 'default'
+kafka.bq.stream.output.mode=<output-mode> default value : append
+```
+
```
Comment (vanshaj-bhatia) [inline java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md:297]: Add description for newly introduced template properties like: `kafka.gcs.batch.interval` `kafka.gcs.consumer.group.id`

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md [hunk 8]
```diff
@@ -237,4 +237,82 @@ bin/start.sh \
...
+--templateProperty kafka.bootstrap.servers=102.1.1.20:9092 \
+--templateProperty kafka.topic=events-topic \
+--templateProperty kafka.starting.offset=latest \
+--templateProperty kafka.bq.stream.output.mode=Append \
+--templateProperty kafka.bq.batch.interval=60000 \
+--templateProperty kafka.bq.consumer.group.id=test.group.id \
+--templateProperty kafka.bq.dataset=kafkatobq \
+--templateProperty kafka.bq.table=kafkaevents \
+--templateProperty kafka.bq.temp.gcs.bucket=<gcs-bucket-name> \
+--templateProperty kafka.bq.await.termination.timeout=<stream-await-termination-timeout>
```
Comment (vanshaj-bhatia) [inline java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md:317]: Add some dummy values in Example Submission. Replace `<gcs-bucket-name>` and `<stream-await-termination-timeout>`:
```
--templateProperty kafka.bq.temp.gcs.bucket=<gcs-bucket-name> \
--templateProperty kafka.bq.await.termination.timeout=<stream-await-termination-timeout>
```

File: java/src/test/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstreamTest.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  private static final Logger LOGGER = LoggerFactory.getLogger(KafkaToBQDstream.class);
+
+  @BeforeEach
+  void setup() {
+    PropertyUtil.getProperties().setProperty(PROJECT_ID_PROP, "projectID");
+    PropertyUtil.getProperties().setProperty(KAFKA_BOOTSTRAP_SERVERS, "some_value");
+    PropertyUtil.getProperties().setProperty(KAFKA_TOPIC, "some_value");
+    PropertyUtil.getProperties().setProperty(KAFKA_BQ_DATASET, "some_value");
+    PropertyUtil.getProperties().setProperty(KAFKA_BQ_TABLE, "some_value");
+    PropertyUtil.getProperties().setProperty(KAFKA_BQ_TEMP_GCS_BUCKET, "some_value");
```
Comment (vanshaj-bhatia) [inline java/src/test/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstreamTest.java:41]: post making KAFKA_BQ_CONSUMER_GROUP_ID as required input, adjust the test case.

**Review by vanshaj-bhatia — CHANGES_REQUESTED** [review]

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstream.java (hunk 4 above)
Comment (chakresh84) [reply java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstream.java:98]: Yes. As per KafkaUtils.createDirectStream ,  it is expecting  list of topics

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md (hunk 8 above)
Comment (chakresh84) [reply java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md]: Done

File: java/pom.xml (hunk 1 above)
Comment (chakresh84) [reply java/pom.xml]: done

File: java/src/main/resources/template.properties (hunk 2 above)
Comment (chakresh84) [reply java/src/main/resources/template.properties]: done

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstream.java (hunk 3 above)
Comment (chakresh84) [reply java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstream.java:153]: done

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstream.java (hunk 5 above)
Comment (chakresh84) [reply java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstream.java]: done

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstream.java (hunk 6 above)
Comment (chakresh84) [reply java/src/main/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstream.java:188]: done

File: java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md (hunk 7 above)
Comment (chakresh84) [reply java/src/main/java/com/google/cloud/dataproc/templates/kafka/README.md:303]: done

File: java/src/test/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstreamTest.java (hunk 9 above)
Comment (chakresh84) [reply java/src/test/java/com/google/cloud/dataproc/templates/kafka/KafkaToBQDstreamTest.java:41]: done

**Comment by vanshaj-bhatia** [comment]
@chakresh84 Please add Integration test to [Java Jenkinsfile](https://github.com/GoogleCloudPlatform/dataproc-templates/blob/main/java/.ci/Jenkinsfile).

ref. https://github.com/GoogleCloudPlatform/dataproc-templates/blob/main/java/.ci/Jenkinsfile#L306

**Comment by vanshaj-bhatia** [comment]
@chakresh84 The integration test appears to be broken. It seems that the Kafka connector is unable to establish a connection with the Kafka Brokers, as evidenced by this ad-hoc pipeline run: [https://dataproctemplatesci.com/blue/organizations/jenkins/dataproc-templates-build%2Fad-hoc%2Fintegration-tests-java/detail/integration-tests-java/404/pipeline/500](https://dataproctemplatesci.com/blue/organizations/jenkins/dataproc-templates-build%2Fad-hoc%2Fintegration-tests-java/detail/integration-tests-java/404/pipeline/500)

Another observation: if the connection isn't successful, the code retries indefinitely, consuming resources unnecessarily. Ideally, timeout should have handled this. Please examine this behavior as well.

**Comment by chakresh84** [comment]
> @chakresh84 The integration test appears to be broken. It seems that the Kafka connector is unable to establish a connection with the Kafka Brokers, as evidenced by this ad-hoc pipeline run: https://dataproctemplatesci.com/blue/organizations/jenkins/dataproc-templates-build%2Fad-hoc%2Fintegration-tests-java/detail/integration-tests-java/404/pipeline/500
> 
@vanshaj-bhatia  @vanshaj-bhatia Yes it failing because it is not below error [1]. Do we know who maintains the kafka server "0.0.0.50:9093" ? Also,  how this is being handled in the other templates which are based on Kafka? 


[1] 
23/12/31 13:33:35 INFO KafkaConsumer: [Consumer clientId=consumer-testgroup-1, groupId=testgroup] Subscribed to topic(s): integration-test
23/12/31 13:33:35 WARN NetworkClient: [Consumer clientId=consumer-testgroup-1, groupId=testgroup] Connection to node -1 (/10.0.0.50:9093) could not be established. Broker may not be available.
23/12/31 13:33:35 WARN NetworkClient: [Consumer clientId=consumer-testgroup-1, groupId=testgroup] Bootstrap broker 10.0.0.50:9093 (id: -1 rack: null) disconnected
23/12/31 13:33:35 WARN NetworkClient: [Consumer clientId=consumer-testgroup-1, groupId=testgroup] Connection to node -1 (/10.0.0.50:9093) could not be established. Broker may not be available.

**Comment by vanshaj-bhatia** [comment]
@chakresh84 I am maintaining Kafka broker "10.0.0.50" on a single VM for integration tests.

I've gone through the logs, debugged the issues, and made adjustments to the VM configuration to enable the connection, and it appears to be working now.

Logs: [Previous](https://dataproctemplatesci.com/blue/rest/organizations/jenkins/pipelines/dataproc-templates-build/pipelines/ad-hoc/pipelines/integration-tests-java/runs/404/nodes/500/steps/510/log/?start=0), [Current](https://dataproctemplatesci.com/blue/organizations/jenkins/dataproc-templates-build%2Fad-hoc%2Fintegration-tests-java/detail/integration-tests-java/406/pipeline/500/)

**Review by vanshaj-bhatia — APPROVED** [review]
Great job! The template look good to me.
Please include this template to both the readme files (root and Java). After merging, proceed with the next step to carry out the hardening of the template.

**Comment by chakresh84** [comment]
> Great job! The template look good to me. Please include this template to both the readme files (root and Java). After merging, proceed with the next step to carry out the hardening of the template.

@vanshaj-bhatia  README changes are done

**Pull request merged**